utilities/transactions/pessimistic_transaction.cc

```diff
@@ -630,7 +630,12 @@ Status PessimisticTransaction::TryLock(ColumnFamilyHandle* column_family,
     // the key is already locked, this func will update some stats on the
     // tracked key. It could also update the tracked_at_seq if it is lower than
     // the existing trackey seq.
-    TrackKey(cfh_id, key_str, tracked_at_seq, read_only, exclusive);
+    //
+    // If assume_tracked is true, we assume that TrackKey has been called
+    // previously.
+    if (!assume_tracked) {
```

Comment:
What if the new lock is exclusive? assumed_tracked definition does not say that? (could be updated though):
```
296   // assume_tracked=true expects the key be already tracked. If valid then it
297   // skips ValidateSnapshot. Returns error otherwise.
```

Reply:
If you're referring to lock upgrades, then I think there might be a minor perf issue with not calling TrackKey.

There's an optimization where we skip calling TryLock if a key is already tracked on `tracked_keys_`, but if we don't update the local map during lock upgrades, I think we'll end up reupgrading the same lock when we don't have to.

Reply:
Although I guess you could argue that assume tracked means, assume this was tracked with the same exclusive settings previously...

Reply:
> assume this was tracked with the same exclusive settings previously...
I let you decide that as you know better how assumed_tracked is used by application.

Reply:
MyRocks doesn't rely on this behaviour, so it's probably a safe assumption to make.